File: server_integration_tests/java/java_server_reactive/src/main/java/dev/graphlink/javaserver/impl/BulkCreateServiceImpl.java
package dev.graphlink.javaserver.impl;

import dev.graphlink.javaserver.generated.inputs.CreateArticleInput;
import dev.graphlink.javaserver.generated.services.BulkCreateService;
import dev.graphlink.javaserver.generated.types.Article;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.List;

@Service
public class BulkCreateServiceImpl implements BulkCreateService {

    @Override
    public Mono<Integer> bulkCreate(List<? extends List<? extends List<CreateArticleInput>>> matrix) {
        if (matrix == null) return Mono.just(0);
        int count = 0;
        for (var group : matrix) {
            if (group == null) continue;
            for (var batch : group) {
                if (batch == null) continue;
                for (var input : batch) {
                    if (input != null) {
                        var article = Article.builder()
                            .id(Data.nextId())
                            .title(input.getTitle())
                            .authorId(input.getAuthorId())
                            .build();
                        Data.articles.add(article);
                        count++;
                    }
                }
            }
        }
        return Mono.just(count);
    }

    @Override
    public Mono<Void> validateBulkCreate(List<? extends List<? extends List<CreateArticleInput>>> matrix) {
        return Mono.empty();
    }
}

File: server_integration_tests/kotlin/kotlin_server_blocking/src/main/kotlin/dev/graphlink/kotlinserverblocking/impl/BulkCreateServiceImpl.kt
package dev.graphlink.kotlinserverblocking.impl

import dev.graphlink.kotlinserverblocking.generated.inputs.CreateArticleInput
import dev.graphlink.kotlinserverblocking.generated.services.BulkCreateService
import dev.graphlink.kotlinserverblocking.generated.types.Article
import org.springframework.stereotype.Service

@Service
class BulkCreateServiceImpl : BulkCreateService {
    override suspend fun bulkCreate(matrix: List<List<List<CreateArticleInput?>>?>?): Int {
        if (matrix == null) return 0
        var count = 0
        for (group in matrix) {
            if (group == null) continue
            for (batch in group) {
                if (batch == null) continue
                for (input in batch) {
                    if (input != null) {
                        Data.articles.add(Article(id = Data.nextId(), title = input.title, authorId = input.authorId))
                        count++
                    }
                }
            }
        }
        return count
    }

    override suspend fun validateBulkCreate(matrix: List<List<List<CreateArticleInput?>>?>?) {}
}
